lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1172 @@
+			// TODO: add more cases of course
+			// TODO: with the llvm class this by default wraps around, do not do that thx
+			llvm::TypeSwitch<mlir::Operation*>(op)
+			.Case<mlir::rlc::AddOp>([&](mlir::rlc::AddOp addop){
+				new_range=result_range.subWithNoWrap(new_range,0,llvm::ConstantRange::PreferredRangeType::Signed);
+			});
+			// Insert the range where is needed
+			return currentLattice->insertOrUpdate(std::make_pair(update,branch),new_range);
+		}
+		// Else case just do nothing
+		return false;
+	}
+
+	// Method called for visiting the yield operation
+	static bool visitYield(
+		mlir::Operation* statement, 
+		mlir::rlc::ConstraintsLattice* currentLattice){
+
+		mlir::Operation* conditional_op=nullptr;
+
+		llvm::TypeSwitch<mlir::Operation*>(statement)
+		.Case<mlir::rlc::Yield>([&](mlir::rlc::Yield op){
+			conditional_op=op.getOperand(0).getDefiningOp();
+		})
+		.Case<mlir::rlc::BuiltinAssignOp>([&](mlir::rlc::BuiltinAssignOp op){
+			conditional_op=op.getOperand(1).getDefiningOp();
+		});
+
+		if(auto casted=mlir::dyn_cast<mlir::rlc::UninitializedConstruct>(conditional_op)){
+			//llvm::outs()<<"\nAdded uninitialized construct:\n";
+			return mlir::rlc::visitUninitializedConstruct(statement, currentLattice);
+		}
+
+		// Do the checking only if the operation is a conditional (relational)
+		if(	not mlir::isa<mlir::rlc::LessOp>(conditional_op) and 
+			not mlir::isa<mlir::rlc::GreaterOp>(conditional_op) and
+			not mlir::isa<mlir::rlc::LessEqualOp>(conditional_op) and 
+			not mlir::isa<mlir::rlc::GreaterEqualOp>(conditional_op))
+			return false;
+
+		// Aliases
+		const auto& lhs = conditional_op->getOperand(0);
+		const auto& rhs = conditional_op->getOperand(1);
+		const auto& lhs_op = lhs.getDefiningOp();
+		const auto& rhs_op = rhs.getDefiningOp();
+
+		// The algorithm will work mostly as the one for the arithmetic operations, but it will be simpler (no commutativity problem)
+		// NB: note that they are not exclusive so they have to be executed always
+
+		auto to_return=false;
+
+		// If the rhs is a constant
+		if(auto rhs_casted=mlir::dyn_cast<mlir::rlc::Constant>(rhs_op)){
+
+			to_return=currentLattice->updateRelationalRange(conditional_op, true, lhs, rhs_casted, true);
+			to_return=currentLattice->updateRelationalRange(conditional_op, false, lhs, rhs_casted, false) or to_return;
+
+			return to_return;
+
+		}
+		// If the lhs is a constant then do the same but with inverted branches
+		else if(auto lhs_casted=mlir::dyn_cast<mlir::rlc::Constant>(lhs_op)){
+
+			to_return=currentLattice->updateRelationalRange(conditional_op, false, rhs, lhs_casted, true);	
+			to_return=currentLattice->updateRelationalRange(conditional_op, true, rhs, lhs_casted, false) or to_return;
+				
+			return to_return;
+		}
+		// else do nothing
+		return false;
+
+	}
+
+	// When I visit an uninitialized construct I need to insert the ssa value 
+	// in the set
+	static bool visitUninitializedConstruct(
```

Comment:
visit unitialized construct does not have much to do with unitialized constructs. Can we rewrite this so that the interface is used instead of specialcasing this guy in particular?